src/map.rs

```diff
@@ -937,6 +937,74 @@ where
         n.leading_zeros() as isize | (1 << (RESIZE_STAMP_BITS - 1)) as isize
     }
 
+    // TODO
+    fn try_presize<'g>(&self, size: usize, guard: &'g Guard) {
+        let new_capacity = if size >= MAXIMUM_CAPACITY / 2 {
+            MAXIMUM_CAPACITY
+        } else {
+            // 1.5 * size + 1
+            let size = size + (size >> 1) + 1;
+
+            std::cmp::min(MAXIMUM_CAPACITY, size.next_power_of_two())
+        };
+
+        let c = new_capacity as isize; // TODO
+
+        loop {
+            let sc = self.size_ctl.load(Ordering::SeqCst);
+            if sc < 0 {
+                break;
+            }
+
+            let table = self.table.load(Ordering::SeqCst, &guard);
+
+            // TODO: clean this up?
+            let n = if table.is_null() {
+                None
+            } else {
+                Some(unsafe { table.deref() }.len())
+            };
+
+            if n.is_none() || n.unwrap() == 0 {
+                let n = sc.max(c) as usize;
+
+                let mut table = self.table.load(Ordering::SeqCst, guard);
+
+                if self.size_ctl.compare_and_swap(sc, -1, Ordering::SeqCst) == sc {
+                    if self.table.load(Ordering::SeqCst, guard) == table {
+                        let new_table = Owned::new(Table::new(n));
+                        table = new_table.into_shared(guard);
+                        self.table.store(table, Ordering::SeqCst);
+                        self.size_ctl
+                            .store(n as isize - (n >> 2) as isize, Ordering::SeqCst);
+                    }
+
+                    self.size_ctl.store(sc, Ordering::SeqCst);
+                }
+            } else if c <= sc || n.unwrap() >= MAXIMUM_CAPACITY {
+                break;
+            } else if table == self.table.load(Ordering::SeqCst, &guard) {
+                let rs: isize = Self::resize_stamp(n.unwrap()) << RESIZE_STAMP_SHIFT;
+                // TODO: `rs` is postive even though `resize_stamp` says:
+                // "Must be negative when shifted left by RESIZE_STAMP_SHIFT"
+                // and since our size_control field needs to be negative
+                // to indicate a resize this needs to be addressed
+
+                if self.size_ctl.compare_and_swap(sc, rs + 2, Ordering::SeqCst) == sc {
+                    self.transfer(table, Shared::null(), &guard);
+                }
+            }
+        }
+    }
+
+    #[inline]
+    /// Tries to reserve capacity for at least additional more elements.
+    /// The collection may reserve more space to avoid frequent reallocations.
+    pub fn reserve(&self, additional: usize) {
+        let guard = epoch::pin();
+        self.try_presize(additional, &guard);
```

Comment:
I _think_ `try_presize` takes the _absolute_ table size as an argument, whereas `reserve` takes `additional`, so this won't work out quite right.

Reply:
Oh right, I totally forgot about that.
But converting from additional to absolute shouldn't be that hard
```rust
let absolute = self.len() + additional;
```
would do it, right?

Reply:
Yes, that should be right. It's racy, but then again, that's essentially what the user requested!